### ISSUE TITLE
OOM when running Gradient Accumulation

### ISSUE THREAD

**Issue by siddk** [pytorch/xla#3593]
## 🐛 Bug

When I write my own (single-core, haven't tested multi-core yet!) loop for running a PyTorch model training with gradient accumulation on TPUs, I get an OOM error when running with gradient accumulation (but no OOM when running without it!). 

I think this is a problem of how the underlying graphs get compiled. Other issues in the repo seem to suggest adding an explicit `xm.mark_step()` but would be good to (1) find out where, and (2) add this information to the docs if possible.

## Sketch of Code

A sketch of my training loop is below:

```python
# bsz = 64

device = xm.xla_device()
model = model.train().to(device)
optimizer = model.configure_optimizer()

# Set accumulation steps --> when this is 1, everything is fine up through a batch size of 128!
accumulate_grad_batches = 8
for train_idx, img in enumerate(train_dataloader):
     img = img.to(device)
     loss = model(img)

      # Normalize loss to account for accumulation
      loss = loss / accumulate_grad_batches
      loss.backward()

       if (train_idx + 1) % accumulate_grad_batches == 0:
           optimizer.step()
           xm.mark_step()
           optimizer.zero_grad()
```

## Expected behavior

Should not OOM with gradient accumulation, as memory footprint should remain the same?

## Environment

 - Reproducible on XLA backend [CPU/TPU]: TPU
 - torch_xla version: 1.11

Running on a TPU-v2-8 VM with the default pt-1.11 image.

**Comment by JackCaoG** [comment]
graph execution won't happen until the `xm.mark_step()` call, everything before that is just graph building. From a quick look with your code, what ended up happening in this case is you are trying to executing a 8x larger graph which caused the oom.

**Comment by siddk** [comment]
Got it - I went ahead and added the `xm.mark_step()` at the end of the loop. For reference, are there docs around when to call `xm.mark_step()` and what it's actually doing?

**Comment by JackCaoG** [comment]
https://github.com/pytorch/xla/blob/master/API_GUIDE.md#xla-tensors-are-lazy should be a good start. You might also want to read https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md and https://pytorch.org/blog/understanding-lazytensor-system-performance-with-pytorch-xla-on-cloud-tpu/

**Comment by Shankar203** [comment]
Hey, what's the right way of doing the gradient accumulation for the SPMD based models?? I couldn't find any doccumentation.
